### ISSUE TITLE
Classes break overload checking

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#11606]
**Bug Report**

A certain overloaded function works, but when put in a class no longer works.

**To Reproduce**

```py
import typing

# this works:
@typing.overload
def f(a: str, /) -> None: ...

@typing.overload
def f(*, b: bool = False) -> None: ...

def f(a: str | None = None, /, *, b: bool = False) -> None:
    ...

# this does not work:
class Blah:
    @typing.overload
    def f(self, a: str, /) -> None: ...
    
    @typing.overload
    def f(self, *, b: bool = False) -> None: ...
    
    def f(self, a: str | None = None, /, *, b: bool = False) -> None:
        ...
```

**Expected Behavior**

For this case, as the `self` argument is not annotated and is always there, mypy shouldn't report any issues.

**Actual Behavior**

```
main.py:21: error: Overloaded function implementation does not accept all possible arguments of signature 2
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

Reproduced in mypy playground: https://mypy-play.net/?mypy=master&python=3.10&flags=strict&gist=d2c0e39b68def2cd02633caeded68e3f

**Comment by sobolevn** [comment]
This happens because for some reason left and right are `def (ex.Blah, Union[builtins.str, None] =, *, b: builtins.bool =)` and `def (self: ex.Blah, *, b: builtins.bool =)`.

It does not pass this check:

https://github.com/python/mypy/blob/b47245adc60eecde9164799c426f3bffcedbc48e/mypy/subtypes.py#L1058-L1077

In contrast to regular functions: `
`def (Union[builtins.str, None] =, *, b: builtins.bool =)` and `def (*, b: builtins.bool =)`.

Notice that `self` is named in classes on the right, but not on the left.
I will try to fix this today! 👍

**Comment by sobolevn** [comment]
This is a corner case of using `/` in a method. Because in this case you make two arguments positional-only, here: `def f(self, a: str, /) -> None: ...` you say: both `self` and `a` are positional only. This is why it is different from functions.

But, another question is: should we support this corner case? 
Because we actually can use `self` as a named argument, when we use `ClassName.method(self=instance)`. It is rare, but a valid case.

When declaring `self` as a positional-only, we break this. So, technically these two signatures are really incompatible.

**Comment by sobolevn** [comment]
Here's the valid code for this case:

```python
class Blah:
    @overload
    def f(self, a: str, /) -> None: ...

    @overload
    def f(self, /, *, b: bool = False) -> None: ...

    def f(self, a: str | None = None, /, *, b: bool = False) -> None:
        ...
```

It works: `Revealed type is "Overload(def (ex.Blah, builtins.str), def (ex.Blah, *, b: builtins.bool =))"`

**Comment by A5rocks** [comment]
> When declaring self as a positional-only, we break this. So, technically these two signatures are really incompatible. @A5rocks, do you agree?

Didn't think of it that way, that is true. But nonetheless I think it should be fixed (it's not *really* a pos argument as long as these are called as methods).
